components/scream/src/share/scream_types.hpp

```diff
@@ -1,18 +1,63 @@
-#ifndef INCLUDE_SCREAM_TYPES
-#define INCLUDE_SCREAM_TYPES
+#ifndef SCREAM_TYPES_HPP
+#define SCREAM_TYPES_HPP
+
+#include <Kokkos_Core.hpp>
 
 #include "share/scream_config.hpp"
 
 namespace scream {
 
 #ifdef SCREAM_DOUBLE_PRECISION
-typedef double Real;
+using Real = double;
 #else
-typedef float Real;
+using Real = float;
 #endif
 
 typedef int Int;
 
+// Detect ExecSpace using some ifdef logic. I put Serial here as a placeholder
+using ExecSpace = Kokkos::Serial;
+using HostSpace = Kokkos::HostSpace;
+
+// Memory spaces
+using ExecMemSpace = ExecSpace::memory_space;
+using HostMemSpace = HostSpace::memory_space;
+
+// Memory traits
+using MemoryManaged   = Kokkos::MemoryTraits<Kokkos::Restrict>;
+using MemoryUnmanaged = Kokkos::MemoryTraits<Kokkos::Restrict | Kokkos::Unmanaged>;
+
+// A general view in Scream has layout right ('row-major'-like)
+template<typename DataType, typename... Properties>
+using ViewType = Kokkos::View<DataType, Kokkos::LayoutRight, Properties...>;
+
+// ======== Some particular view's aliases ========== //
+
+// Managed/Unmanaged view
+template <typename DataType, typename... Properties>
+using ViewManaged = ViewType<DataType, Properties..., MemoryManaged>;
+template <typename DataType, typename... Properties>
+using ViewUnmanaged = ViewType<DataType, Properties..., MemoryUnmanaged>;
+
+// Host/Deviceviews
+template <typename DataType, typename... Properties>
+using HostView = ViewType<DataType, HostMemSpace, Properties...>;
+template <typename DataType, typename... Properties>
+using ExecView = ViewType<DataType, ExecMemSpace, Properties...>;
```

Comment:
We need to plan for mixed execution spaces, such as OpenMP/Cuda. In P3, we're avoiding these top-level type aliases in favor of type dictionaries templated on the device. So all of these top-level aliases go into a type dictionary.

Reply:
Sounds reasonable. I mostly copied and pasted HOMMEXX stuff, but I guess hybrid parallelism is going to happen in the lifetime of scream.